File: macosx/AboutWindowController.m
/******************************************************************************
 * $Id$
 *
 * Copyright (c) 2007-2012 Transmission authors and contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *****************************************************************************/

#import "AboutWindowController.h"

#include "version.h"

@implementation AboutWindowController

AboutWindowController * fAboutBoxInstance = nil;
+ (AboutWindowController *) aboutController
{
    if (!fAboutBoxInstance)
        fAboutBoxInstance = [[self alloc] initWithWindowNibName: @"AboutWindow"];
    return fAboutBoxInstance;
}

- (void) awakeFromNib
{
    [fVersionField setStringValue: @(LONG_VERSION_STRING)];
    
    [fCopyrightField setStringValue: [[NSBundle mainBundle] localizedStringForKey: @"NSHumanReadableCopyright"
                                        value: nil table: @"InfoPlist"]];
    
    [[fTextView textStorage] setAttributedString: [[[NSAttributedString alloc] initWithPath:
            [[NSBundle mainBundle] pathForResource: @"Credits" ofType: @"rtf"] documentAttributes: nil] autorelease]];
    
    //size license button
    const CGFloat oldButtonWidth = NSWidth([fLicenseButton frame]);
    
    [fLicenseButton setTitle: NSLocalizedString(@"License", "About window -> license button")];
    [fLicenseButton sizeToFit];
    
    NSRect buttonFrame = [fLicenseButton frame];
    buttonFrame.size.width += 10.0;
    buttonFrame.origin.x -= NSWidth(buttonFrame) - oldButtonWidth;
    [fLicenseButton setFrame: buttonFrame];
}

- (void) windowDidLoad
{
    [[self window] center];
}

- (void) windowWillClose: (id) sender
{
	[fAboutBoxInstance autorelease];
    fAboutBoxInstance = nil;
}

- (IBAction) showLicense: (id) sender
{
    NSString * licenseText = [NSString stringWithContentsOfFile: [[NSBundle mainBundle] pathForResource: @"COPYING" ofType: nil]
                                usedEncoding: nil error: NULL];
    [fLicenseView setString: licenseText];
    [fLicenseCloseButton setTitle: NSLocalizedString(@"OK", "About window -> license close button")];
	
	[NSApp beginSheet: fLicenseSheet modalForWindow: [self window] modalDelegate: nil didEndSelector: nil contextInfo: nil];
}

- (IBAction) hideLicense: (id) sender
{
    [fLicenseSheet orderOut: nil];
    [NSApp endSheet: fLicenseSheet];
}

@end

File: macosx/AboutWindowController.swift
// This file Copyright © 2007-2023 Transmission authors and contributors.
// It may be used under the MIT (SPDX: MIT) license.
// License text can be found in the licenses/ folder.

import AppKit

class AboutWindowController: NSWindowController {
    private static var fAboutWindowInstance: AboutWindowController?
    @objc static var aboutController: AboutWindowController {
        if let fAboutWindowInstance = fAboutWindowInstance {
            return fAboutWindowInstance
        }
        let aboutWindowInstance = AboutWindowController(windowNibName: "AboutWindow")
        self.fAboutWindowInstance = aboutWindowInstance
        return aboutWindowInstance
    }

    @IBOutlet private weak var fTextView: NSTextView!
    @IBOutlet private weak var fLicenseView: NSTextView!
    @IBOutlet private weak var fVersionField: NSTextField!
    @IBOutlet private weak var fCopyrightField: NSTextField!
    @IBOutlet private weak var fLicenseButton: NSButton!
    @IBOutlet private weak var fLicenseCloseButton: NSButton!
    @IBOutlet private weak var fLicenseSheet: NSPanel!

    override func awakeFromNib() {
        super.awakeFromNib()
        fVersionField.stringValue = LONG_VERSION_STRING

        fCopyrightField.stringValue = Bundle.main.localizedString(forKey: "NSHumanReadableCopyright", value: nil, table: "InfoPlist")

        // swiftlint:disable:next force_try
        let credits = try! NSAttributedString(url: Bundle.main.url(forResource: "Credits", withExtension: "rtf")!,
                                              options: [ .documentType: NSAttributedString.DocumentType.rtf ],
                                              documentAttributes: nil)
        fTextView.textStorage?.setAttributedString(credits)

        // size license button
        let oldButtonWidth = fLicenseButton.frame.width

        fLicenseButton.title = NSLocalizedString("License", comment: "About window -> license button")
        fLicenseButton.sizeToFit()

        var buttonFrame = fLicenseButton.frame
        buttonFrame.size.width += 10.0
        buttonFrame.origin.x -= buttonFrame.width - oldButtonWidth
        fLicenseButton.frame = buttonFrame
    }

    override func windowDidLoad() {
        window?.center()
    }

    @IBAction private func showLicense(_ sender: Any) {
        // swiftlint:disable:next force_try
        fLicenseView.string = try! String(contentsOfFile: Bundle.main.path(forResource: "COPYING", ofType: nil)!)
        fLicenseCloseButton.title = NSLocalizedString("OK", comment: "About window -> license close button")

        window?.beginSheet(fLicenseSheet, completionHandler: nil)
    }

    @IBAction private func hideLicense(_ sender: Any) {
        window?.endSheet(fLicenseSheet)
    }
}

extension AboutWindowController: NSWindowDelegate {
    func windowWillClose(_ notification: Notification) {
        Self.fAboutWindowInstance = nil
    }
}
